packages/controllers/src/services/AbstractExecutionService.ts

```diff
@@ -77,12 +85,38 @@ export abstract class AbstractExecutionService<JobType extends Job>
    *
    * @param jobId - The id of the job to be terminated.
    */
-  public terminate(jobId: string): void {
+  public async terminate(jobId: string): Promise<void> {
     const jobWrapper = this.jobs.get(jobId);
     if (!jobWrapper) {
       throw new Error(`Job with id "${jobId}" not found.`);
     }
 
+    let timeout: number | undefined;
+
+    const timeoutPromise = new Promise<void>((resolve) => {
+      timeout = setTimeout(() => {
+        // No need to reject here, we just resolve and move on if the terminate request doesn't respond quickly
+        resolve();
+      }, this._terminationTimeout) as unknown as number;
+    });
+
+    // Ping worker and tell it to run teardown, continue with termination if it takes too long
+    try {
+      await Promise.race([
+        this._command(jobId, {
+          jsonrpc: '2.0',
+          method: 'terminate',
+          params: [],
+          id: nanoid(),
+        }),
+        timeoutPromise,
+      ]);
+    } catch {
+      // Ignore any potential errors, just keep terminating
```

Comment:
Should we log the error here?

Reply:
We can, but it's not an error we care about since the worker is being terminated anyway right?

Reply:
The show will gone on regardless, but I'd like to see how it plays out in practice. I'll add a `console.warning` and we can remove it if it becomes annoying or is in fact useless.